### PR TITLE
Add lock mechanism while submitting and killing LSF jobs

### LINKED ISSUE #7907: Killing LSF jobs while submission is in progress can leave the job running on the cluster.
If an ensemble experiment is started, and 'Terminate experiment' is clicked while realizations are being submitted, the following "Error" is displayed to the user:

https://github.com/equinor/ert/blob/401d29f04aef827f27011482f42cf6c6a153e8fb/src/ert/scheduler/lsf_driver.py#L289-L291

image

This means that the killing of the realization failed. When the `submit()` function eventually finished, the realization is likely to be running even if it should have been killed.

### PR DESCRIPTION
**Pull request by berland**
**Issue**
Resolves #7907 

**Approach**
🔒 

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [x] Make sure tests pass locally (after every commit!)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [x] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/scheduler/lsf_driver.py [hunk 1]
```diff
@@ -227,6 +229,11 @@ async def submit(
         name: str = "dummy",
         runpath: Optional[Path] = None,
     ) -> None:
+        if iens not in self._submit_locks:
+            self._submit_locks[iens] = asyncio.Lock()
+
+        await self._submit_locks[iens].acquire()
```
Comment (xjules) [inline src/ert/scheduler/lsf_driver.py:235]: when looking at the code I would prefer to use the:
```python
async with self._submit_locks[iens]:
 ....
```
which then automatically releases the lock

Comment (berland) [reply src/ert/scheduler/lsf_driver.py:235]: I sort of agree, but I dislike very long indented blocks changing so many lines of code..

**Comment by berland** [comment]
Testing the lock mechanism is blocked by #8020

File: src/ert/scheduler/lsf_driver.py (hunk 1 above)
Comment (berland) [reply src/ert/scheduler/lsf_driver.py:235]: Using async.with now.

Comment (xjules) [reply src/ert/scheduler/lsf_driver.py]: I think it is required to use the `async with Lock` as this one guarantees to release the lock when exception occurs :+1:

Comment (berland) [reply src/ert/scheduler/lsf_driver.py]: fixed.

File: tests/integration_tests/scheduler/test_lsf_driver.py [hunk 2]
```diff
@@ -217,3 +218,47 @@ def mock_poll_once_by_bhist(*args, **kwargs):
...
+    # Allow submit and kill to be interleaved by asyncio by issuing
+    # submit() in its own asyncio Task:
+    asyncio.create_task(
+        driver.submit(
+            # The sleep is the time window in which we can kill the job before
+            # the unwanted finish message appears on disk.
+            0,
+            "sh",
+            "-c",
+            f"sleep 1; touch {tmp_path}/survived",
```
Comment (xjules) [inline tests/integration_tests/scheduler/test_lsf_driver.py:249]: Shall we sleep (eg 5 secs) more here? Just to remove potential flakiness....

Comment (berland) [reply tests/integration_tests/scheduler/test_lsf_driver.py]: Added some if's to guard against flakyness. This test is also relevant in interactive pytest runs so then I don't want to sleep for too long.

File: tests/integration_tests/scheduler/test_lsf_driver.py [hunk 3]
```diff
@@ -217,3 +219,60 @@ def mock_poll_once_by_bhist(*args, **kwargs):
...
+
+    async def finished(iens: int, returncode: int):
+        SIGTERM = 15
+        assert iens == 0
+        assert returncode == SIGNAL_OFFSET + SIGTERM
+
+    await poll(driver, {0}, finished=finished)
+
+    assert "ERROR" not in str(caplog.text)
+    await asyncio.sleep(test_grace_time)
```
Comment (xjules) [inline tests/integration_tests/scheduler/test_lsf_driver.py:277]: Why do we need to sleep here?

Comment (berland) [reply tests/integration_tests/scheduler/test_lsf_driver.py:277]: We have in the `poll()` await ensured that the realization has been reported back to the scheduler with  a correct error code. But, if that is erroneous, and the shell script that is running is not killed (it is the mother process of the shell script that is actually killed), then the `survived` file would still appear on disk. So we wait until we know the survived file would have been written and then assert it is not there.

Comment (xjules) [reply tests/integration_tests/scheduler/test_lsf_driver.py:277]: Right, but that deserves a comment before the `await asyncio.sleep(test_grace_time)` line.

Comment (berland) [reply tests/integration_tests/scheduler/test_lsf_driver.py:282]: added.

**Review by xjules — APPROVED** [review]
Very nice PR @berland ! :rocket:

**Pull request merged**